keps/sig-apps/0032-portable-service-definitions.md

```diff
@@ -0,0 +1,143 @@
+
+3rd party controllers implementing the CRDs can use a variety of methods to do those including using the Kubernetes Service Catalog and leveraging vendor specific APIs.
+
+To work out the details we will start with one service (e.g., MySQL) and go through the process to make is work as managed by Kubernetes and each of the top three public clouds as defined by adoption. Public clouds and other cloud platforms following the top three are welcome to be involved in the process but are not required in the process.
+
+Service providers that implement the services can have a variety of features and settings that are specific to them. Along with that organizations can have policies they want used by default as their settings. To facilitate default settings, especially those specific to a provider, the controllers that acts on the CRs can, optionally, have cluster and namespace configuration passed to them providing the default settings.
+
+### Risks and Mitigations
+
+Two known risks include:
+
+1. The access controls and relationships between accounts and services. How will proper user and tenancy information be passed to clouds that require this form of information?
+2. Details required, when requesting a service, can vary between cloud providers that will implement this as a SaaS. How can that information be commonized or otherwise handled?
+
+## Graduation Criteria
+
+The following are the graduation criteria:
+
+- X organizations have adopted using the portable service definitions
+- Service definitions for at least 3 services have been created and are in use
+- Documented exists explaining how a controller implementer can use the CRDs and secrets to create a controller of their own
+- Service consumer documentation exists explaining how to use portable service definitions within their applications
+
+## Implementation History
+
+- the _Summary_, _Motivation_, _Proposal_ sections developed
+
+## Drawbacks [optional]
+
+Why should this KEP _not_ be implemented.
+
+## Alternatives [optional]
+
+An alternative is to modify the service catalog to leverage CRDs and return common secret credentials. Drawbacks to this are that it would be a complicated re-write to the service catalog, according to the service catalog team, and the solution would still require the open service broker (OSB) to be implemented in all environments (e.g., Minikube) even where simpler models (e.g., a controller) could be used instead. The solution proposed here could work with the service catalog in environments it makes sense and use other models in other environments. The focus here is more on the application operator experience working with services than all of the implementations required to power it.
+
```

Comment:
We have experienced a need a similar CRD object to connect two controllers or controller with a SAAS service. Currently we are using something called `AppBinding` based on service catalog's `ServiceBinding` https://github.com/kmodules/custom-resources/blob/master/apis/appcatalog/v1alpha1/appbinding_types.go#L20

I would like to see service-catalog to move to CRD and just use the `ServiceBinding` object instead. I think that already captures the necessary information for "connecting" applications.

Reply:
@tamalsaha My initial thought was this would be a good route for service catalog. After investigating it I came to a different opinion. Here are some things I learned.

1. Service catalog is tightly coupled to the Open Service Broker (OSB). It's even noted in their charter. Not every environment needs OSB and some folks would like to try a different route (e.g., [AWS](https://github.com/awslabs/aws-service-operator)). Why does this work need to be coupled with the OSB if some folks want to take a different route? Can we have a solution that can work with OSB/Service Catalog and other systems?
2. The changes needed to move the service catalog to a CRD based system like this are substantial according to the people I spoke with who work on it. Service catalog folks started working on this idea this way before I even started talking with them. It just needed a home outside of their SIG because of the scope.
3. There is momentum to do this now and trying to figure out how to alter the service catalog will likely kill the momentum, from what I can tell.

It's worth noting that code to move forward on this in this manner was started by people who are currently active on the service catalog. The demo in sig apps was from a sig service catalog co-chair.